specs/merkle-tree.md

```diff
@@ -0,0 +1,142 @@
+* `string uri`: An optional field that contains the URI of the deployment. Its purpose is to provide a public record that allows anyone to re-assemble the entire deployment from scratch. This may include the Solidity compiler inputs, which are required for Etherscan verification. The format, location, and contents of the URI are determined by off-chain tooling.
+* `bool arbitraryChain`: If this is `true`, the Merkle root can be executed on any chain without the explicit permission of the Gnosis Safe owners. This is useful if the owners want their system to be permissionlessly deployed on new chains. By default, this is disabled, which means that the Gnosis Safe owners must explicitly approve the deployment on individual chains.
+
+### `EXECUTE` Leaf Data
+
+The `data` field of an `EXECUTE` leaf primarily contains data to forward to the Gnosis Safe. It consists of the following fields, which are ABI encoded:
+* `address to`: The target address of the transaction. This is _not_ the address of the Gnosis Safe.
+* `uint256 value`: The amount of native gas token to transfer from the Gnosis Safe to the target address. This value is not transferred from the `SphinxModuleProxy`.
+* `uint256 gas`: The amount of gas that's included in the call from the `SphinxModuleProxy` to the Gnosis Safe.
+* `bytes txData`: The transaction's data.
+* `Enum.Operation operation`: The type of transaction to execute in the Gnosis Safe, i.e. `Call` or `DelegateCall`.
+* `bool requireSuccess`: If this is `true` and the transaction in the Gnosis Safe fails, the deployment is marked as "failed" and will end immediately. If this is `false`, the deployment will continue regardless of whether the transaction fails.
+
+### `CANCEL` Leaf Data
+TODO
+
+## Valid Merkle Trees
+There are a wide variety of trees that conform to the basic structure described above, but not all of them will necessarily be executable on-chain for all networks they contain leaves for. For example, it is possible to include two leaves with the same `index` and `chainId`. In that case, both leaves would be valid but only one would be executable.
+
+To resolve these types of problems we outline a set of Merkle tree invariants below that define what a 'valid' Merkle tree looks like. We expect relayers to rely on Merkle trees following the invariants set out here. We assume that relayers will refuse to execute Merkle trees that do not follow these invariants.
+
+## High-Level Merkle Tree Invariants
+
+### 1. Must be executable on all chains for which there is at least one Merkle leaf.
+Assumption: The input Merkle leaf(s) data satisfies the on-chain conditions in the corresponding SphinxModuleProxy contract(s). See the [SphinxModuleProxy specification](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/specs/sphinx-module-proxy.md#function-level-invariants) for more information.
```

Comment:
I think we should link to the assumptions section, where we discuss this in slightly more detail, instead of the SphinxModuleProxy spec